ols/user_data_collection/data_collector.py

```diff
@@ -0,0 +1,303 @@
+        f"Unknown value in INGRESS_ENV: {INGRESS_ENV}. Allowed: prod, stage"
+    )
+INGRESS_TIMEOUT = 10  # seconds
+INGRESS_BASE_DELAY = 60  # exponential backoff parameter
+INGRESS_MAX_RETRIES = 3  # exponential backoff parameter
+
+CP_OFFLINE_TOKEN = os.environ.get("CP_OFFLINE_TOKEN")
+REDHAT_SSO_TIMEOUT = 5  # seconds
+
+# TODO: OLS-473
+# OLS_USER_DATA_MAX_SIZE = 100 * 1024 * 1024  # 100 MB
+
+if INGRESS_ENV == "stage" and not CP_OFFLINE_TOKEN:
+    raise ValueError("CP_OFFLINE_TOKEN is required for stage environment")
+
+
+logging.basicConfig(
+    level=logging.DEBUG,
+    stream=sys.stdout,
+    format="[%(asctime)s] %(levelname)s: %(message)s",
+)
+# silence libs
+# - urllib3 - we don't care about those debug posts
+# - kubernetes - prints resources content when debug, causing secrets leak
+logging.getLogger("kubernetes").setLevel(logging.WARNING)
+logging.getLogger("urllib3").setLevel(logging.WARNING)
+
+logger = logging.getLogger(__name__)
+
+
+def get_ingress_upload_url() -> str:
+    """Get the Ingress upload URL based on the environment."""
+    upload_endpoint = "api/ingress/v1/upload"
+    if INGRESS_ENV == "prod":
+        return "https://console.redhat.com/" + upload_endpoint
+    elif INGRESS_ENV == "stage":
+        return "https://console.stage.redhat.com/" + upload_endpoint
+    else:
+        raise ValueError(f"Unknown value in INGRESS_ENV: {INGRESS_ENV}")
+
+
+def access_token_from_offline_token(offline_token: str) -> str:
+    """Generate "access token" from the "offline token".
+
+    Offline token can be generated at:
+        prod - https://access.redhat.com/management/api
+        stage - https://access.stage.redhat.com/management/api
+
+    Args:
+        offline_token: Offline token from the Customer Portal.
+
+    Returns:
+        Refresh token.
+    """
+    if INGRESS_ENV == "stage":
+        url = "https://sso.stage.redhat.com/"
+    else:
+        url = "https://sso.redhat.com/"
+    endpoint = "auth/realms/redhat-external/protocol/openid-connect/token"
+    data = {
+        "grant_type": "refresh_token",
+        "client_id": "rhsm-api",
+        "refresh_token": offline_token,
+    }
+
+    response = requests.post(url + endpoint, data=data, timeout=REDHAT_SSO_TIMEOUT)
+    if response.status_code == requests.codes.ok:
+        return response.json()["access_token"]
+    else:
+        raise Exception(f"Failed to generate access token. Response: {response.json()}")
+
+
+def get_cloud_openshift_pull_secret() -> str:
+    """Get the pull secret token from the cluster."""
+    kubernetes.config.load_incluster_config()
+    v1 = kubernetes.client.CoreV1Api()
+
+    # get the pull-secret from openshift-config namespace
+    secret = v1.read_namespaced_secret("pull-secret", "openshift-config")
+    dockerconfigjson = secret.data[".dockerconfigjson"]
+    dockerconfig = json.loads(base64.b64decode(dockerconfigjson).decode("utf-8"))
+
+    return dockerconfig["auths"]["cloud.openshift.com"]["auth"]
```

Comment:
There is a story for that https://issues.redhat.com/browse/OLS-500

Reply:
that follow up story is good, but this code still needs to handle the situation in the meantime, otherwise we're going to get ugly exception/errors thrown.

Reply:
Right. The operator should ensure the secret is there before the container is deployed, so I've added a temp solution for this.

Reply:
> I've added a temp solution for this.

what solution did you add? i don't see the scenario handled here.

Reply:
Latest commits?

Reply:
ah, you handled it elsewhere.  i'd have just expected you to handle the possibility that the value isn't found here in this logic.

among other things it's also possible that at startup the value doesn't exist, but later it is added.  or vice versa.  So better to check every time (and properly handle that the entry might be missing) than to do it once at startup.

not to mention the crashloop issue w/ your current startup checking logic.